x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/queries/SemanticQueryBuilder.java

```diff
@@ -0,0 +1,270 @@
+            return this;
+        }
+
+        Set<String> modelsForField = queryRewriteContext.getModelsForField(fieldName);
+        if (modelsForField.isEmpty()) {
+            throw new IllegalArgumentException("Field [" + fieldName + "] is not a semantic_text field type");
+        }
+
+        if (modelsForField.size() > 1) {
+            // TODO: Handle multi-index semantic queries
+            throw new IllegalArgumentException("Field [" + fieldName + "] has multiple models associated with it");
+        }
+
+        InferenceAction.Request inferenceRequest = new InferenceAction.Request(
+            TaskType.ANY,
+            modelsForField.iterator().next(),
+            List.of(query),
+            Map.of(),
+            InputType.SEARCH
+        );
+
+        SetOnce<InferenceServiceResults> inferenceResultsSupplier = new SetOnce<>();
+        queryRewriteContext.registerAsyncAction((client, listener) -> executeAsyncWithOrigin(
+            client,
+            ML_ORIGIN,
+            InferenceAction.INSTANCE,
+            inferenceRequest,
+            listener.delegateFailureAndWrap((l, inferenceResponse) -> {
+                inferenceResultsSupplier.set(inferenceResponse.getResults());
+                l.onResponse(null);
+            })
+        ));
+
+        return new SemanticQueryBuilder(this, inferenceResultsSupplier);
+    }
+
+    @Override
+    protected Query doToQuery(SearchExecutionContext context) throws IOException {
+        InferenceServiceResults inferenceServiceResults = inferenceResultsSupplier.get();
+        if (inferenceServiceResults == null) {
+            throw new IllegalArgumentException("Inference results supplier for field [" + fieldName + "] is empty");
+        }
+
+        List<? extends InferenceResults> inferenceResultsList = inferenceServiceResults.transformToCoordinationFormat();
+        if (inferenceResultsList.isEmpty()) {
+            throw new IllegalArgumentException("No inference results retrieved for field [" + fieldName + "]");
+        } else if (inferenceResultsList.size() > 1) {
+            // TODO: How to handle multiple inference results?
+            throw new IllegalArgumentException(inferenceResultsList.size() + " inference results retrieved for field [" + fieldName + "]");
+        }
+
+        InferenceResults inferenceResults = inferenceResultsList.get(0);
+        MappedFieldType fieldType = context.getFieldType(fieldName);
+        if (fieldType instanceof SemanticTextFieldMapper.SemanticTextFieldType == false) {
+            // TODO: Better exception type to throw here?
+            throw new IllegalArgumentException(
+                "Field [" + fieldName + "] is not registered as a " + SemanticTextFieldMapper.CONTENT_TYPE + " field type"
+            );
+        }
+
+        return semanticQuery(inferenceResults, context);
+    }
+
+    private Query semanticQuery(InferenceResults inferenceResults, SearchExecutionContext context) {
```

Comment:
Moved the `semanticQuery` method here so that query generation logic is encapsulated in a single class.

Reply:
Ok, I'm fine with either approach. I went with this approach because all the examples I found encapsulated the query logic in `QueryBuilder`.

Reply:
@jimczi any guidance on whether `semanticQuery` belongs here or in `SemanticTextFieldMapper`?

I put it here because `SemanticTextFieldMapper` doesn't contain any logic related to the "shadow" nested mappings used to index inference results, all of that is in `SemanticTextInferenceResultFieldMapper`. My thought was that putting `semanticQuery` in this class encapsulates the logic related to querying the "shadow" nested mappings in one class, much like how `SemanticTextInferenceResultFieldMapper` encapsulates the logic for indexing inference results.

Reply:
I agree with @carlosdelest , having the query logic on the field type also allows to test things when the field is configured through the mapping instead of re-creating the fields outside.

Reply:
Done 👍